File: scripts/gh_admin_mod/features/concurrency.py
from gh_admin_mod.env import get_env, parse_bool
from gh_admin_mod.github_api import count_open_items
from gh_admin_mod.models import FeatureResult, ModerationContext
from gh_admin_mod.templates import render_template


def _parse_limits(raw_limits: str) -> dict[str, int]:
    limits: dict[str, int] = {}
    for part in raw_limits.split(","):
        if "=" not in part:
            continue
        key, val = part.split("=", 1)
        try:
            limits[key.strip().upper()] = int(val.strip())
        except ValueError:
            continue
    return limits


def evaluate(context: ModerationContext) -> FeatureResult:
    if not parse_bool(get_env("INPUT_CONCURRENCY_ENABLED", "false")):
        return FeatureResult(matched=False)

    raw_limits = get_env(
        "INPUT_CONCURRENCY_LIMITS",
        "NONE=1,FIRST_TIMER=1,FIRST_TIME_CONTRIBUTOR=1,CONTRIBUTOR=10,COLLABORATOR=10,MEMBER=10",
    )
    limits = _parse_limits(raw_limits)

    limit = limits.get(context.author_association.upper())

    # If limit is 0, it means unlimited. If None, no limit defined for this association.
    if limit is None or limit <= 0:
        return FeatureResult(matched=False)

    current_count = count_open_items(context, context.author, context.item_type)

    if current_count <= limit:
        return FeatureResult(matched=False)

    result = FeatureResult(
        matched=True,
        feature="concurrency",
        reason=(
            f"@{context.author} has {current_count} open {context.item_type_plural}, "
            f"exceeding the limit of {limit} for {context.author_association}."
        ),
        metadata={
            "limit": str(limit),
            "count": str(current_count),
        },
    )

    default_comment = (
        "This {type} was automatically closed because @{user} already has "
        "{count} open {type_plural}, exceeding the allowed limit of {limit}."
    )
    comment_template = get_env("INPUT_CONCURRENCY_COMMENT_MESSAGE", default_comment)
    result.comment_message = render_template(comment_template, context, result)

    return result

File: src/features/concurrency.ts
import * as core from '@actions/core';
import { parseBool, parseIntSafe } from '../env.js';
import { countOpenItems } from '../github_api.js';
import { FeatureResult, ModerationContext } from '../models.js';
import { renderTemplate } from '../templates.js';

function parseLimits(rawLimits: string): Record<string, number> {
  const limits: Record<string, number> = {};
  for (const part of rawLimits.split(',')) {
    if (!part.includes('=')) continue;
    const [key, val] = part.split('=', 2);
    const parsedVal = parseInt(val.trim(), 10);
    if (!isNaN(parsedVal)) {
      limits[key.trim().toUpperCase()] = parsedVal;
    }
  }
  return limits;
}

export async function evaluate(context: ModerationContext): Promise<FeatureResult> {
  if (!parseBool(core.getInput('concurrency-enabled') || 'false')) {
    return {
      matched: false,
      feature: '',
      reason: '',
      labels: [],
      lock_conversation: false,
      metadata: {},
    };
  }

  const rawLimits =
    core.getInput('concurrency-limits') ||
    'NONE=1,FIRST_TIMER=1,FIRST_TIME_CONTRIBUTOR=1,CONTRIBUTOR=10,COLLABORATOR=10,MEMBER=10';
  const limits = parseLimits(rawLimits);

  const limit = limits[context.author_association.toUpperCase()];

  // If limit is 0, it means unlimited. If undefined, no limit defined for this association.
  if (limit === undefined || limit <= 0) {
    return {
      matched: false,
      feature: '',
      reason: '',
      labels: [],
      lock_conversation: false,
      metadata: {},
    };
  }

  const currentCount = await countOpenItems(context, context.author, context.item_type);

  if (currentCount <= limit) {
    return {
      matched: false,
      feature: '',
      reason: '',
      labels: [],
      lock_conversation: false,
      metadata: {},
    };
  }

  const result: FeatureResult = {
    matched: true,
    feature: 'concurrency',
    reason:
      `@${context.author} has ${currentCount} open ${context.item_type_plural}, ` +
      `exceeding the limit of ${limit} for ${context.author_association}.`,
    labels: [],
    lock_conversation: false,
    metadata: {
      limit: limit.toString(),
      count: currentCount.toString(),
    },
  };

  const defaultComment =
    'This {type} was automatically closed because @{user} already has ' +
    '{count} open {type_plural}, exceeding the allowed limit of {limit}.';
  const commentTemplate = core.getInput('concurrency-comment-message') || defaultComment;
  result.comment_message = renderTemplate(commentTemplate, context, result);

  return result;
}
